src/digit/mod.rs

```diff
@@ -56,27 +57,311 @@ pub trait Digit: AsRef<str> {
+    "دو" => 2,
+    "سه" => 3,
+    "چهار" => 4,
+    "پنج" => 5,
+    "شش" => 6,
+    "هفت" => 7,
+    "هشت" => 8,
+    "نه" => 9,
+    "ده" => 10,
+    "یازده" => 11,
+    "دوازده" => 12,
+    "سیزده" => 13,
+    "چهارده" => 14,
+    "پانزده" => 15,
+    "شانزده" => 16,
+    "هفده" => 17,
+    "هجده" => 18,
+    "نانزده" => 19,
+    "بیست" => 20,
+    "سی" => 30,
+    "چهل" => 40,
+    "پنجاه" => 50,
+    "شصت" => 60,
+    "هفتاد" => 70,
+    "هشتاد" => 80,
+    "نود" => 90,
+    "صد" => 100,
+    "دویست" => 200,
+    "سیصد" => 300,
+    "جهارصد" => 400,
+    "پانصد" => 500,
+    "ششصد" => 600,
+    "هفتصد" => 700,
+    "هشتصد" => 800,
+    "نهصد" => 900,
+};
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+enum TokenType {
+    Multiplier(u32),
+    FaceValue(u16),
+}
+
+fn parse(token: &str) -> Result<TokenType, &'static str> {
+    if let Some(v) = FACE_VALUE.get(token) {
+        Ok(TokenType::FaceValue(*v))
+    } else if let Some(m) = MULTIPLIERS.get(token) {
+        Ok(TokenType::Multiplier(*m))
+    } else {
+        Err("Unsupported token")
+    }
+}
+
+/// Extension trait for conversion of number strings in written format into numbers.
+///
+/// See [`words_to_number`] for more details.
+pub trait WordsToNumber: AsRef<str> {
+    /// Convert `&self` into a number with given type `N`.
+    ///
+    /// Needless to say, the resulting number must fit into `N`.
+    ///
+    /// Supports only values from [0, 999_999_999_999].
+    ///
+    /// The given string may contain only tokens being equal to either of [`FACE_VALUE`] or
+    /// [`MULTIPLIERS`], or the special case of "و". Any other token will lead to a parse error.
+    // IMPLEMENTATION NOTE: we follow the semantics of a basic stack-machine: Any face-value type
+    // will be accumulated, until a multiplier is seen, at which point, it is multiplied into the
+    // accumulated face-values, and we reset.
+    fn words_to_number<
+        N: num_traits::Zero
+            + num_traits::One
+            + num_traits::CheckedMul
+            + num_traits::CheckedAdd
+            + std::convert::TryFrom<u32>
+            + std::convert::TryFrom<u16>
+            + std::fmt::Debug
+            + Copy,
+    >(
+        &self,
+    ) -> Result<N, &'static str> {
+        // TODO: ^^ maybe make a module-level Result alias.
+        use std::convert::TryInto;
+        const CANT_CONVERT: &'static str = "Given number does not fit in the provided`N`";
+
+        let src = self.as_ref().trim();
+        let tokens = src.split(" ").collect::<Vec<_>>();
+
+        let mut last: Option<TokenType> = None;
+        let parsed = tokens
+            .into_iter()
+            .filter(|t| (*t).ne("و"))
+            .map(parse)
+            .collect::<Result<Vec<_>, _>>()?;
+
+        let mut final_value: N = num_traits::Zero::zero();
+        let mut intermediary_value: N = num_traits::Zero::zero();
+
+        let check_add = |this: N, that: N| this.checked_add(&that).ok_or(CANT_CONVERT);
```

Comment:
Do we really need to check for an overflow for addition and multiplication?
Currently, the maximum number can be 999_999_999_999.
Seems like it's not possible to have an overflow.

Reply:
Ah, OK, N is coming from the function call.
Maybe we can just use i128 instead?